cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1592,36 +1593,54 @@ void reader::impl::allocate_columns(read_mode mode, size_t skip_rows, size_t num
     auto const d_cols_info = cudf::detail::make_device_uvector_async(
       h_cols_info, _stream, cudf::get_current_device_resource_ref());
 
-    auto const num_keys = _input_columns.size() * max_depth * subpass.pages.size();
-    // size iterator. indexes pages by sorted order
-    rmm::device_uvector<size_type> size_input{num_keys, _stream};
-    thrust::transform(
-      rmm::exec_policy(_stream),
-      thrust::make_counting_iterator<size_type>(0),
-      thrust::make_counting_iterator<size_type>(num_keys),
-      size_input.begin(),
-      get_page_nesting_size{
-        d_cols_info.data(), max_depth, subpass.pages.size(), subpass.pages.device_begin()});
-    auto const reduction_keys =
-      cudf::detail::make_counting_transform_iterator(0, get_reduction_key{subpass.pages.size()});
     cudf::detail::hostdevice_vector<size_t> sizes{_input_columns.size() * max_depth, _stream};
-
-    // find the size of each column
-    thrust::reduce_by_key(rmm::exec_policy(_stream),
-                          reduction_keys,
-                          reduction_keys + num_keys,
-                          size_input.cbegin(),
-                          thrust::make_discard_iterator(),
-                          sizes.d_begin());
-
-    // for nested hierarchies, compute per-page start offset
-    thrust::exclusive_scan_by_key(
-      rmm::exec_policy(_stream),
-      reduction_keys,
-      reduction_keys + num_keys,
-      size_input.cbegin(),
-      start_offset_output_iterator{
-        subpass.pages.device_begin(), 0, d_cols_info.data(), max_depth, subpass.pages.size()});
+    auto const num_keys              = _input_columns.size() * max_depth * subpass.pages.size();
+    auto constexpr max_keys_per_iter = static_cast<size_t>(
+      std::numeric_limits<size_type>::max() / 2);  ///< Maximum 1billion keys per iteration
+    auto const num_keys_per_iter =
+      num_keys < max_keys_per_iter
+        ? num_keys
+        : (max_depth * subpass.pages.size()) *
+            std::max<size_t>(1, max_keys_per_iter / (max_depth * subpass.pages.size()));
```

Comment:
Why not simply `std::min<size_t>(max_depth * subpass.pages.size(), max_keys_per_iter)`?

Reply:
`std::max<size_t>(1, max_keys_per_iter / (max_depth * subpass.pages.size()))` portion essentially calculates how many input cols we can do in each iteration. Total keys are `ncols x max depth x total pages` and we want to cut along the `ncols` axis if that makes sense?

Reply:
So essentially we need `variable x max_depth x total_pages)` where `variable` is `>= 1` and is the largest number which would produce a product smaller than the threshold.

Reply:
Apologies, I messed up when I interpreted the code.
How about a `num_keys_per_column`? replaces `max_depth * subpass.pages.size()`, which appears at least three times.

Reply:
> essentially calculates how many max input cols we can do in an iteration

I think this can be the way it's calculated as well (have `max_cols_per_iter`, etc.), so we don't have to transform the code to translate it into our mental model.

Not a blocking suggestion